### ISSUE TITLE
A 10 line script takes 40 seconds to transpile

### ISSUE THREAD

**Issue by nchanged** [Microsoft/TypeScript#18005]
Hi!
We use typescript to transpile some es6 modules. I recently found a very annoying piece of code in `moment`  library that just hangs typescript for 40 (!!!!) seconds

**TypeScript Version:**  `2.6.0-dev.20170824`

**Code**

```js
const ts = require("typescript")
const start = new Date().getTime()
let result = ts.transpileModule(`
let seconds, minutes, hours = 0;
let thresholds = {}
var a = seconds <= thresholds.ss && ['s', seconds] ||
    seconds < thresholds.s && ['ss', seconds] ||
    minutes <= 1 && ['m'] ||
    minutes < thresholds.m && ['mm', minutes] ||
    hours <= 1 && ['h'] ||
    hours < thresholds.h && ['hh', hours] ||
    days <= 1 && ['d'] ||
    days < thresholds.d && ['dd', days] ||
    months <= 1 && ['M'] ||
    months < thresholds.M && ['MM', months] ||
    years <= 1 && ['y'] || ['yy', years];

`, {
    compilerOptions: {
        module: "es6"
    }
});
const end = new Date().getTime();
console.log(result);
console.log("Took", (end - start));
```

I played around with it for a bit, and discovered that removing `years <= 1 && ['y'] || ['yy', years];`  (the last line) makes it compile pretty much instantly.

It's easy to reproduce: copy the snippet, create a js file (or ts) doesn't matter, and execute it.

Thanks a lot!

**Comment by grantila** [comment]
```ts
var a =
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4];
```
took 4.5 minutes to compile.

Seems both
```ts
var a =
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    .
    .
    .;
```
and
```ts
var a =
    x && [1, 2, ...] ||
    x && [1, 2, ...] ||
    x && [1, 2, ...] ||
    x && [1, 2, ...];
```
individually causes exponential compile time. Combining them obviously causes exponential^2 compile times.

I get the same result on 2.4.

**Comment by ahejlsberg** [comment]
Fixed in #18174.

Until the fix ships, a simple workaround is to add a type annotation to the variable being initialized, e.g.

```ts
var a: number[] =
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4] ||
    x && [1, 2, 3, 4];
```

**Comment by nchanged** [comment]
👍
